### PR TITLE
Add MLX LoRA training and audio encoding primitives

### PR DESCRIPTION
**Pull request by betweentwomidnights**
## Summary

This adds pure-MLX primitives for training and applying Stable Audio 3 LoRA-family adapters within the existing `optimized/mlx` runtime.
These mirror the adapter types already supported by the official PyTorch implementation.

- Trainable MLX `Linear` and `Conv1d` adapters
- LoRA, DoRA rows/columns, BoRA, and their XS variants
- SA3-compatible safetensors loading and saving
- Canonical checkpoint-name mapping for the optimized MLX model layout
- PyTorch-to-MLX `Conv1d` weight-layout conversion
- Rectified-flow loss, training timestep samplers, and distribution shifts
- SAME waveform-to-latent encoding, including codec alignment, overlapping
  chunking, and latent validity masks
- Fixed-strength inference application for one or more checkpoints

The optimized runtime remains MLX/NumPy-only. PyTorch and the separate `safetensors` package are not added as runtime dependencies.

## Motivation

These primitives were developed while adding Apple Silicon LoRA training to the open-source gary4local application. This PR adapts that downstream work to the official optimized model definitions and checkpoint conventions rather than importing the application-specific training pipeline.

The longer downstream implementation and adaptation record is available here for optional context:

https://github.com/betweentwomidnights/gary-localhost-installer-mac/blob/main/docs/sa3/STABLE_AUDIO_3_UPSTREAM_PR_NOTES.md

## Checkpoint compatibility

Saved checkpoints use the existing parametrization-style tensor keys and `lora_config` metadata. Tests load MLX-generated checkpoints through the official PyTorch loader and compare every supported adapter variant directly against the official PyTorch adapter math.

The implementation also handles two optimized-runtime compatibility details:

- `to_local_embed.0/2` checkpoint names map to the optimized
  `to_local_embed.seq.0/2` modules.
- PyTorch `Conv1d` weights use `[out, in, kernel]`, while MLX uses
  `[out, kernel, in]`.

## Audio pre-encoding

The MLX training loss operates on SAME latents. `encode_audio()` supplies the required waveform-to-latent boundary through the existing optimized SAME-S and SAME-L encoders. It pads waveforms to the codec's required alignment, supports overlapping chunked encoding for longer clips, and returns a ceiling-scaled validity mask for padded batches.

`encode_audio()` begins after audio has been decoded and resampled. It accepts 44.1 kHz stereo waveforms; dataset-level concerns such as file discovery, captions, and saving encoded latents remain with the calling training workflow.

## Inference scope

`apply_lora_checkpoint()` and `apply_lora_checkpoints()` materialize requested strengths into a loaded model's in-memory weights. They do not modify the base checkpoint on disk.

This is intended for fixed-strength, load-time inference. It supports multiple checkpoints with independent strengths, but it is deliberately not presented as a persistent slider API: applying a new strength repeatedly to the same model instance would compound adapted weights.

Because Stable Audio 3 LoRAs are especially expressive when blended together as a set of adjustable strengths, as the existing Gradio UI supports for PyTorch users, a follow-up PR is planned around a reusable MLX generation
pipeline and a base-preserving adapter session. That layer will support persistent inference, multiple simultaneously loaded LoRAs with independently adjustable strengths, and MLX integration for the CLI/Gradio interface.

## Cross-backend listening tests

A matched ear test applied the same downstream DoRA checkpoint in this optimized runtime and in Gary4local using the same prompt, seed, sampling steps, and adapter strength. The outputs were audibly different, but both had the expected adapter character and comparable quality; subjectively, they sounded like two pieces from the same song.

Waveform identity is not expected between the two pipelines. For the tested duration, the optimized runtime generates 87 latent positions while gary4local aligns to 88 and supplies a padding mask. The pipelines also consume MLX random state in a different order. A shared numeric seed only produces identical noise when random calls and tensor shapes also match, and the extra latent position changes both the sampled tensor shape and model attention context.

## Deliberately out of scope

- Dataset discovery, file decoding/resampling, and latent-cache persistence
- A complete optimizer/training command
- Changes to the existing CLI or Gradio interface
- Application-specific defaults, job management, or UI

Those pieces can consume these primitives without making this initial API review depend on a larger interface refactor.

## Validation

- `uvx ruff check .`
- `uvx ruff format --check .`
- `26 passed` across the focused MLX audio-encoding, LoRA, and training tests
- Existing official PyTorch LoRA test passes
- Real optimized medium-DiT forward, backward, and AdamW smoke step
- Real SAME-L encoding of a 16-second stereo clip exactly matches the existing
  optimized inference path; chunked encoding differs by at most `5.5e-4`
- A 500-step rank-4 DoRA run using the proposed encoder, adapter, timestep,
  loss, optimizer, and checkpoint primitives completed with finite loss on
  every step; mean loss moved from `0.792` over the first 50 steps to `0.495`
  over the final 50
- The resulting 108-tensor checkpoint loads through the official PyTorch
  checkpoint reader and applies to all 36 expected optimized ARC targets with
  no missing or skipped layers
- In a listening test of that newly trained checkpoint, a bell-arpeggio source
  trained with the `garybell` trigger audibly introduced its bell character
  into a lo-fi hip-hop generation using the same trigger. This provides a
  perceptual end-to-end check of pre-encoding, prompt conditioning, training,
  checkpoint loading, and inference application.
- Separately, an existing 500-step DoRA checkpoint produced by Gary4local
  applied to all 36 expected optimized medium layers with no missing or
  skipped targets, providing a backwards-compatibility check for downstream
  checkpoints created before this upstream adaptation.

Together, these checks cover both directions of the intended boundary:
training a new adapter with the proposed official MLX primitives, and loading an existing downstream adapter into the official optimized runtime.

### CODE REVIEW

File: optimized/mlx/models/defs/audio_encoding.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+            f"pad_modulo={pad_modulo}."
+        )
+
+    total_latents = total_patches // encoder_stride
+    if not chunked or total_latents <= chunk_size:
+        return encoder(patches)
+    if chunk_size < 1:
+        raise ValueError("chunk_size must be positive.")
+    if overlap < 0 or overlap >= chunk_size:
+        raise ValueError("overlap must be non-negative and smaller than chunk_size.")
```
[the remark this answers is not in the document]
Comment (betweentwomidnights) [reply optimized/mlx/models/defs/audio_encoding.py:188]: Interesting catch, thanks! I reproduced the odd-overlap path and confirmed that it does not drop or duplicate latent positions: `_stitch_encoded_chunks()` clips each interval's `target_end` to the next interval's actual start and clips `target_start` against the output cursor. For odd overlaps, the extra latent is intentionally assigned to the later chunk.

I added an explanatory comment and a regression test covering both even and odd configured overlaps, including an end-anchored final chunk with a larger odd actual overlap. Chunked output matches unchunked output exactly in that test. I also swept small total lengths, chunk sizes, and all valid even/odd overlaps locally with exact matches.

Out of curiosity, we examined the downstream training pipeline in gary4local and it had the same potential ambiguity.

**Comment by Cortexelus** [comment]
integrated into #72
